checker/src/org/checkerframework/checker/lock/LockStore.java

```diff
@@ -137,4 +159,83 @@ protected void internalDotOutput(StringBuilder result) {
+    public void updateForMethodCall(MethodInvocationNode n,
+            AnnotatedTypeFactory atypeFactory, CFValue val) {
+        ExecutableElement method = n.getTarget().getMethod();
+
+        // case 1: remove information if necessary
+        if (!(((LockChecker) analysis.getTypeFactory().getContext()).hasOption("assumeSideEffectFree")
+              || isSideEffectFree(atypeFactory, method))) {
+            // update field values
+            Map<FlowExpressions.FieldAccess, CFValue> newFieldValues = new HashMap<>();
+            for (Entry<FlowExpressions.FieldAccess, CFValue> e : fieldValues.entrySet()) {
+                FlowExpressions.FieldAccess fieldAccess = e.getKey();
+                CFValue otherVal = e.getValue();
+
+                // case 3:
+                List<Pair<AnnotationMirror, AnnotationMirror>> fieldAnnotations = atypeFactory
+                        .getAnnotationWithMetaAnnotation(
+                                fieldAccess.getField(),
+                                MonotonicQualifier.class);
+                CFValue newOtherVal = null;
+                for (Pair<AnnotationMirror, AnnotationMirror> fieldAnnotation : fieldAnnotations) {
+                    AnnotationMirror monotonicAnnotation = fieldAnnotation.second;
+                    Name annotation = AnnotationUtils.getElementValueClassName(
+                            monotonicAnnotation, "value", false);
+                    AnnotationMirror target = AnnotationUtils.fromName(
+                            atypeFactory.getElementUtils(), annotation);
+                    AnnotationMirror anno = otherVal.getType()
+                            .getAnnotationInHierarchy(target);
+                    // Make sure the 'target' annotation is present.
+                    if (anno != null && AnnotationUtils.areSame(anno, target)) {
+                        newOtherVal = analysis.createSingleAnnotationValue(
+                                target, otherVal.getType().getUnderlyingType())
+                                .mostSpecific(newOtherVal, null);
+                    }
+                }
+                if (newOtherVal != null) {
+                    // keep information for all hierarchies where we had a
+                    // monotone annotation.
+                    newFieldValues.put(fieldAccess, newOtherVal);
+                    continue;
+                }
+
+                // case 2:
+                if (!fieldAccess.isUnmodifiableByOtherCode()) {
+                    continue; // remove information completely
+                }
+
+                // keep information
+                newFieldValues.put(fieldAccess, otherVal);
+            }
+            fieldValues = newFieldValues;
+
+            // Necessary because a method could unlock a lock that is a local variable, e.g.:
+            // ReentrantLock lock = new ReentrantLock();
+            // lock.lock();
+            // unlockMyLock(lock);
+            localVariableValues.clear();
```

Comment:
I will try this approach. It shouldn't matter if the local variables are cleared before or after the replaceValue.